crates/ruff_python_formatter/src/statement/suite.rs

```diff
@@ -490,6 +503,40 @@ fn stub_file_empty_lines(
     }
 }
 
+/// Checks if an empty line should be inserted between the preceding and, optionally,
+/// the following node according to the [`blank_line_after_nested_stub_class`](https://github.com/astral-sh/ruff/issues/8891)
+/// preview style.
+///
+/// If `following` is `None`, then the preceding node is the last node in the suite.
+pub(crate) fn blank_line_after_nested_stub_class_condition(
+    preceding: AnyNodeRef<'_>,
+    following: Option<AnyNodeRef<'_>>,
+    f: &PyFormatter,
+) -> bool {
+    let comments = f.context().comments();
+    match preceding.as_stmt_class_def() {
+        Some(class) if contains_only_an_ellipsis(&class.body, comments) => {
+            !class.decorator_list.is_empty()
+                || match following {
+                    Some(AnyNodeRef::StmtClassDef(ast::StmtClassDef {
+                        body,
+                        decorator_list,
+                        ..
+                    })) => !contains_only_an_ellipsis(body, comments) || !decorator_list.is_empty(),
+                    Some(AnyNodeRef::StmtFunctionDef(_)) | None => true,
+                    _ => false,
+                }
+        }
+        Some(_) => !comments.has_trailing_own_line(preceding),
+        None => std::iter::successors(
+            preceding.last_child_in_body(),
+            AnyNodeRef::last_child_in_body,
+        )
+        .take_while(|last_child| !comments.has_trailing_own_line(*last_child))
+        .any(|last_child| last_child.is_stmt_class_def()),
```

Comment:
If the preceding node isn't a class, then we check if there's a class node as the last child node of the body of nested nodes. This nesting could be arbitrarily deep.